rbd: have dummy attacher implementation

humblec opened the pull request:
previously, it was a requirement to have attacher sidecar for CSI
    drivers and there had an implementation of dummy mode of operation.
    However skipAttach implementation has been stabilized and the dummy
    mode of operation is going to be removed from the external-attacher.
    Considering this driver  work on volumeattachment objects for NBD driver
    use cases, we have to implement dummy controllerpublish and unpublish
    and thus keep supporting our operations even in absence of dummy mode
    of operation in the sidecar.
    
    This commit make a NOOP controller publish and unpublish for RBD driver.
    
    CephFS driver does not require attacher and it has already been made free
    from the attachment operations.
    
    Ref# https://github.com/ceph/ceph-csi/pull/3149
    Ref# https://github.com/kubernetes-csi/external-attacher/issues/226

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

nixpanic commented:
Why?

humblec commented:
/retest all

nixpanic commented:
> /retest all

Please don't do this for PRs that are not completely ready, or do not require exhaustive testing yet. The CI environment is rather busy already, re-testing like this just prevents other PRs from getting merged sooner.

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
@nixpanic added the details to the commit and PR. Also tests are passing ! ptal.. thanks..

nixpanic commented on internal/cephfs/controllerserver.go:
```diff
@@ -1066,3 +1066,19 @@ func deleteSnapshotAndUndoReservation(
 	return nil
 }
+
+// ControllerPublishVolume publish volume on node.
```
Why include this in CephFS if it does not need it? Keep it in the `DefaultControllerServer` instead.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1559,3 +1559,35 @@ func (cs *ControllerServer) ControllerExpandVolume(
+	ctx context.Context,
+	req *csi.ControllerPublishVolumeRequest,
+) (*csi.ControllerPublishVolumeResponse, error) {
+	// A dummy publish implementation to mimic a successful attach operation being a NOOP
```
a description of the function should be above the function, that counts for `ControllerUnpublishVolume` too.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1559,3 +1559,35 @@ func (cs *ControllerServer) ControllerExpandVolume(
+	}
+
+	return &csi.ControllerPublishVolumeResponse{
+		PublishContext: map[string]string{},
```
add a comment why this map is (or needs to be) empty, please

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -1559,3 +1559,35 @@ func (cs *ControllerServer) ControllerExpandVolume(
+	req *csi.ControllerPublishVolumeRequest,
+) (*csi.ControllerPublishVolumeResponse, error) {
+	// A dummy publish implementation to mimic a successful attach operation being a NOOP
+	if req.VolumeId == "" {
```
use functions like `req.GetVolumeId()` instead of accessing the attribute directly.

humblec committed:
rbd: add dummy attacher implementation

previously, it was a requirement to have attacher sidecar for CSI
drivers and there had an implementation of dummy mode of operation.
However skipAttach implementation has been stabilized and the dummy
mode of operation is going to be removed from the external-attacher.
Considering this driver  work on volumeattachment objects for NBD driver
use cases, we have to implement dummy controllerpublish and unpublish
and thus keep supporting our operations even in absence of dummy mode
of operation in the sidecar.

This commit make a NOOP controller publish and unpublish for RBD driver.

CephFS driver does not require attacher and it has already been made free
from the attachment operations.

    Ref# https://github.com/ceph/ceph-csi/pull/3149
    Ref# https://github.com/kubernetes-csi/external-attacher/issues/226

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
All the test failures are on

```
Errors during downloading metadata for repository 'tcmu-runner':

  - Status code: 404 for https://3.chacra.ceph.com/r/tcmu-runner/master/245914c1446dddf07d5b58b0a7b2060b50fde4d7/centos/8/flavors/default/x86_64/repodata/repomd.xml (IP: 158.69.93.173)

Error: Failed to download metadata for repo 'tcmu-runner': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried

Error: error building at STEP "RUN dnf -y install 	librados-devel librbd-devel 	/usr/bin/cc 	make 	git 	&& dnf clean all 	&& rm -rf /var/cache/yum     && true": error while running runtime: exit status 1

make: *** [Makefile:233: image-cephcsi] Error 1

script returned exit code 2
```

humblec commented:
@Mergifyio rebase

humblec commented:
@nixpanic comments are addressed.. ptal.. thanks

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.23

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

humblec commented:
@nixpanic can you revisit as I have addressed the changes requested here ?

nixpanic commented:
Is there a reason to keep `ControllerPublishVolume` in `internal/csi-common/controllerserver-default.go`? Or does it make sense to move this dummy implementation there instead?

If `ControllerPublishVolume` is only needed for NBD support, it should probably be reported as a feature only for NBD backed volumes. But, I guess the controller does not report the capabilities depending on the parameters for the volume?

humblec commented:
> Is there a reason to keep `ControllerPublishVolume` in `internal/csi-common/controllerserver-default.go`? Or does it make sense to move this dummy implementation there instead?

in this case, it default to not `implemented` ( inherited for cephfs) and only get an implementation in case of RBD. 
> 
> If `ControllerPublishVolume` is only needed for NBD support, it should probably be reported as a feature only for NBD backed volumes. But, I guess the controller does not report the capabilities depending on the parameters for the volume?

Yeah, this return fall into get caps call for the controller..

humblec commented:
@nixpanic I have answered the queries.. ptal. thanks.

humblec commented:
@nixpanic ptal

humblec commented:
@nixpanic ptal. Thanks

pkalever commented:
@humblec is this a preparation for removing the `registry.k8s.io/sig-storage/csi-attacher` sidecar from RBD provisioner in the future?

humblec commented:
> @humblec is this a preparation for removing the `registry.k8s.io/sig-storage/csi-attacher` sidecar from RBD provisioner in the future?

This is not related to that @pkalever

humblec commented:
@Mergifyio refresh

humblec commented:
@Mergifyio rebase

pkalever commented:
> > @humblec is this a preparation for removing the `registry.k8s.io/sig-storage/csi-attacher` sidecar from RBD provisioner in the future?
> 
> This is not related to that @pkalever

Then I'm confused.

> previously, it was a requirement to have attacher sidecar for CSI
> drivers and there had an implementation of dummy mode of operation.

Well even now we have the csi-attacher sidecar, right?
Don't know what I'm missing here, could you please help me understand?
